### PR TITLE
specs: Ecotone / 4844 / Dencun

### PR DESCRIPTION
**Pull request by protolambda**
**Description**

This PR introduces EIP-4844 and L2-Dencun changes as draft for the Ecotone upgrade.

There are some placeholder specifications still, which I'd like to collaborate on with everyone involved in the design / implementation of these features:
- [x] The blob-encoding (@roberto-bayardo)
- [x] `L1Block` contract-upgrading (@danyalprout)
- [x] `L1Block` contract changes (@danyalprout @anikaraghu)
- [x] L1-data-fees specifications (@roberto-bayardo @K-Ho )

### CODE REVIEW

File: specs/derivation.md
Comment (ajsutton) [inline specs/derivation.md]: If this is done as a deposit transaction it will be included before any L2 tx in that block and they may be unable to access correct L1 price data. If the contract upgrade is backwards compatible and can return the current pre-blob L1 data after the upgrade that's fine - otherwise the upgrade may cause issues for L2 transactions included in the last block before the tx.

And more generally, it feels very weird to activate an upgrade the block before it activates. Sadly the L1Info tx really has to be the first in the block because a heap of code in various places has baked in that dependency (shame we didn't think of upgrade tx earlier but such is life).
If the L1Info deposit tx is unchanged I'd do the contract upgrade in the activation block (after the L1Info tx). If the L1Info tx is different, I think I'd use the old deposit tx in the activation block, then the upgrade tx and the new deposit tx format is used from activation block + 1. That's ugly but ensures no changes happen until the activation block and it's inline with how the dao fork was implemented across multiple blocks so it's not unprecedented.


I

File: specs/superchain-upgrades.md
Comment (ajsutton) [inline specs/superchain-upgrades.md]: ?
```suggestion
- Deneb (Consensus Layer): *not applicable to L2*
```

**Review by ajsutton — COMMENTED** [review]
Looks good generally - I'm a bit concerned about upgrading the L1 info contract a block early but there aren't any great transactions so I could be convinced.

Given the L1 attributes predeployed contract is changing, we probably should update the [L1 Attributes Deposited Transaction](https://github.com/ethereum-optimism/optimism/blob/9ff3ebb3983be52c3ca189423ae7b4aec94e0fde/specs/deposits.md#l1-attributes-deposited-transaction) section to note the different call data that's required.  The definition is actually very generic but currently points to the pre-Eclipse contract and could easily be misleading if we don't at least note that the contract call changes as part of Eclipse activation.

File: specs/derivation.md
Comment (trianglesphere) [reply specs/derivation.md]: I like the idea of an upgrade transaction, but in this case I don't love that we break the L1 Info contract for a block. It'd be possible but a lot of work to allow upgrade transactions before the L1 Info tx (ok to do b/c it's a HF).

File: specs/derivation.md
Comment (EvanJRichard) [inline specs/derivation.md]:
```suggestion
#### Ecotone: Blob Retrieval

With the Ecotone upgrade the retrieval stage is extended to support an additional DA source: [EIP-4844] blobs.
Upon each traversed transaction in the batch inbox:
```

File: specs/derivation.md
Comment (EvanJRichard) [inline specs/derivation.md]:
```suggestion
- `sourceHash`: `0x7795a90486cc207315616d57bdbe5ca0ad63b22b5ba7fe087d11774f5de6e10b`,
  computed with the "Upgrade-deposited" type, with `intent = "Ecotone: L1Block upgrade"`
```

File: specs/derivation.md
Comment (EvanJRichard) [inline specs/derivation.md]:
```suggestion
```bash
# compute intent hash:
cast keccak "Ecotone: L1Block upgrade"
# 0xaf2b20ee05be9fc3f0712050591a5f8988f94b56cdf48842863a773b76634fde

# source hash type:
# 0x0000000000000000000000000000000000000000000000000000000000000002

# compute source hash:
cast keccak 0x0000000000000000000000000000000000000000000000000000000000000002af2b20ee05be9fc3f0712050591a5f8988f94b56cdf48842863a773b76634fde
# 0x7795a90486cc207315616d57bdbe5ca0ad63b22b5ba7fe087d11774f5de6e10b
```
```

File: specs/derivation.md
Comment (EvanJRichard) [inline specs/derivation.md]:
```suggestion
#### Ecotone: Beacon block roots contract deployment (EIP-4788)
```

File: specs/derivation.md
Comment (EvanJRichard) [inline specs/derivation.md]:
```suggestion
```bash
# compute intent hash:
cast keccak "Ecotone: beacon block roots contract deployment"
# 0xab0dfc96b47739a0ae1d415bbfaae79ebb1111861a3b7cfbbaa6ca4a9e618357

# source hash type:
# 0x0000000000000000000000000000000000000000000000000000000000000002

# compute source hash:
cast keccak 0x0000000000000000000000000000000000000000000000000000000000000002ab0dfc96b47739a0ae1d415bbfaae79ebb1111861a3b7cfbbaa6ca4a9e618357
# 0x69b763c48478b9dc2f65ada09b3d92133ec592ea715ec65ad6e7f3dc519dc00c
```
```

File: specs/derivation.md
Comment (EvanJRichard) [inline specs/derivation.md]:
```suggestion
- `sourceHash`: `0x69b763c48478b9dc2f65ada09b3d92133ec592ea715ec65ad6e7f3dc519dc00c`,
```

File: specs/derivation.md
Comment (tynes) [reply specs/derivation.md]: Note that this part of the spec is modified in https://github.com/ethereum-optimism/optimism/pull/8702 and it is out of date here until that PR is merged

File: specs/deposits.md
Comment (tynes) [inline specs/deposits.md]: This section is out of date

**Comment by tynes** [comment]
> Looks good generally - I'm a bit concerned about upgrading the L1 info contract a block early but there aren't any great transactions so I could be convinced.

We have changed the functionality here such that the hf block contains the contract upgrade and then the next block contains the new call

File: specs/deposits.md
Comment (danyalprout) [reply specs/deposits.md]: I'll put in a PR to update this 🙏

File: specs/predeploys.md
Comment (protolambda) [inline specs/predeploys.md:238]: TODO: need to clarify that these revert after the ecotone upgrade.
The `decimals` value is hardcoded, but I think it should also revert, since we don't use it anymore.

File: specs/deposits.md
Comment (maurelian) [inline specs/deposits.md:263]: Nit: this name feels a bit clunky. 

Changing the arguments will allow this to exist alongside the existing function, is the `Ecotone` postfix desirable?

Comment (roberto-bayardo) [reply specs/deposits.md:263]: The thinking here was to make it clear the functions are in fact independent and post-Ecotone upgrade you must call the  new one.

Comment (maurelian) [reply specs/deposits.md:263]: Sounds reasonable. I won't block on this, but I would be a little bit worried about having a lot of this upgrade name post-fix scar tissue kicking around though.
I'd be interested in hear thoughts from other reviewers as they drive by. :)

Comment (tynes) [reply specs/deposits.md:263]: The idea is that we postfix everything with the hardfork name starting with bedrock. This gives a very clear and forwards compatible way of doing things across the whole codebase. fwiw, the geth codebase follows a similar pattern

File: specs/derivation.md
Comment (tynes) [inline specs/derivation.md:1076]: We need to specify the first `setL1BlockValuesEcotone` call to hold the invariant that the view of the system config matches what is on L1. Per https://github.com/ethereum-optimism/optimism/pull/8723, the current scalar value should be migrated to the `baseFeeScalar` and the `blobBaseFeeScalar` should be 0

File: specs/superchain-upgrades.md
Comment (tynes) [inline specs/superchain-upgrades.md:327]: This is underspecified - when you say not applicable to L2, does that mean it will always return 0?

File: specs/system_config.md
Comment (tynes) [inline specs/system_config.md:106]: Perhaps it would be good to clarify that the evm will still pad to 32 bytes so the values will be in the least significant bytes tightly packed together in 8 bytes. It could be good to clarify the byte offsets of the values and their endianness

Comment (tynes) [reply specs/system_config.md:106]: What is the rationale for using `encodePacked` when it still pads to 32 bytes? It makes the parsing logic a bit different for this handler compared to the others and future versions of `solc` want to remove `encodePacked`. It would still be possible to tightly pack them without that helper function if we did want to upgrade but just wondering since it doesn't save any data size

Comment (roberto-bayardo) [reply specs/system_config.md:106]: my understanding was if we used regular abi.Encode it would be 32 bytes per uint for 64 bytes total.

Comment (tynes) [reply specs/system_config.md:106]: ah you are correct

File: specs/superchain-upgrades.md
Comment (protolambda) [reply specs/superchain-upgrades.md:327]: My intention was to leave it functional, but let it return what it would have returned if there was never any blob tx included. We should probably explain in the spec that it then actually return `1`, not `0`. When running `CalcBlobFee(excessBlobGas=0)` that's what you get in geth.

Comment (tynes) [reply specs/superchain-upgrades.md:327]: https://github.com/ethereum-optimism/optimism/pull/8830

**Review by tynes — APPROVED** [review]

File: specs/predeploys.md
Comment (protolambda) [reply specs/predeploys.md:238]: tracked here now: https://github.com/ethereum-optimism/protocol-quest/issues/73

**Pull request merged**
